src/primitives/hrpstring.rs

```diff
@@ -0,0 +1,447 @@
+            has_upper = true;
+        } else if ch.is_ascii_lowercase() {
+            has_lower = true;
+        }
+    }
+    if has_upper && has_lower {
+        Err(Error::MixedCase)
+    } else if let Some(pos) = sep_pos {
+        Ok(pos)
+    } else {
+        Err(Error::MissingSeparator)
+    }
+}
+
+/// Errors types for Bech32 (hrpstring) encoding / decoding.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum Error {
+    /// Human-readable part is invalid.
+    InvalidHrp(hrp::Error),
+    /// String does not contain the separator character.
+    MissingSeparator,
+    /// No characters after the separator.
+    NothingAfterSeparator,
+    /// Attempt conversion of an invalid witness version string/number.
+    InvalidWitnessVersion(gf32::Error),
+    /// The data payload is not a valid length.
+    InvalidDataLength,
+    /// The checksum does not match the rest of the data.
+    InvalidChecksum,
+    /// The checksum is not a valid length.
+    InvalidChecksumLength,
+    /// Some part of the string contains an invalid character.
+    InvalidChar(char),
+    /// The bit conversion failed due to a padding issue.
+    InvalidPadding,
+    /// The whole string must be of one case.
+    MixedCase,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match *self {
+            InvalidHrp(ref e) => write_err!(f, "invalid human-readable part"; e),
+            MissingSeparator => write!(f, "missing human-readable separator, \"{}\"", SEP),
+            NothingAfterSeparator => write!(f, "invalid data - no characters after the separator"),
+            InvalidWitnessVersion(ref e) => write_err!(f, "witness version error"; e),
+            InvalidDataLength => write!(f, "invalid data - payload is not a valid length"),
+            InvalidChecksum => write!(f, "invalid checksum"),
+            InvalidChecksumLength => write!(f, "the checksum is not a valid length"),
+            InvalidChar(n) => write!(f, "invalid character (code={})", n),
+            InvalidPadding => write!(f, "invalid padding"),
+            MixedCase => write!(f, "mixed-case strings not allowed"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            InvalidHrp(ref e) => Some(e),
+            InvalidWitnessVersion(ref e) => Some(e),
+            MissingSeparator
+            | NothingAfterSeparator
+            | InvalidDataLength
+            | InvalidChecksum
+            | InvalidChecksumLength
+            | InvalidChar(_)
+            | InvalidPadding
+            | MixedCase => None,
+        }
+    }
+}
+
+impl From<hrp::Error> for Error {
+    fn from(e: hrp::Error) -> Self { Error::InvalidHrp(e) }
+}
+
+impl From<Infallible> for Error {
+    fn from(v: Infallible) -> Self { match v {} }
+}
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

I like this but I'm a bit curious where it comes from. We don't actually use `Infallible` anywhere. Is this a "best practice" thing or is it maybe a holdover from an earlier version of this project that actually did use it?

In the interest of API stability I'm tempted to drop these conversions and revisit them whenever `!` is stabilized.

Reply:
I can't remember why I added this, the compiler lead me to the original need for it and I copied the idea from some code Kix wrote somewhere IIRC. Its not used now so I've removed it.